hil/ext/switches/dellnos9.py

```diff
@@ -0,0 +1,377 @@
+            return None
+        begin = response.find('NativeVlanId:') + len('NativeVlanId:')
+        end = response.find('.', begin)
+        vlan = response[begin:end]
+
+        # At this point, we are unable to remove the default native vlan.
+        # The switchport defaults to native vlan 1. Our deployment tests make
+        # assertions that a switchport has no default vlan at start.
+        # To get the tests to pass, we return None if we see the switchport has
+        # a default native vlan (=1). This needs to be fixed ASAP.
+        if vlan == '1':
+            return None
+
+        return ('vlan/native', vlan)
+
+    def _get_port_info(self, interface):
+        """Returns the output of a show interface command. This removes all
+        spaces from the response before returning it which is then parsed by
+        the caller.
+
+        Sample Response:
+        u"<outputxmlns='http://www.dell.com/ns/dell:0.1/root'>\n<command>
+        showinterfacesswitchportGigabitEthernet1/3\r\n\r\nCodes:U-Untagged,
+        T-Tagged\r\nx-Dot1xuntagged,X-Dot1xtagged\r\nG-GVRPtagged,M-Trunk\r\n
+        i-Internaluntagged,I-Internaltagged,v-VLTuntagged,V-VLTtagged\r\n\r\n
+        Name:GigabitEthernet1/3\r\n802.1QTagged:Hybrid\r\nVlanmembership:\r\n
+        QVlans\r\nU1512\r\n T1511 \r\n\r\n NativeVlanId:1512 .\r\n\r\n\r\n\r\n
+        MOC-Dell-S3048-ON#</command>\n</output>\n"
+        """
+
+        command = 'interfaces switchport %s %s' % \
+            (self.interface_type, interface)
+        response = self._execute(interface, SHOW, command)
+        return response.text.replace(' ', '')
+
+    def _add_vlan_to_trunk(self, interface, vlan):
+        """ Add a vlan to a trunk port.
+
+        If the port is not trunked, its mode will be set to trunk.
+
+        Args:
+            interface: interface to add the vlan to
+            vlan: vlan to add
+        """
+        if not self._is_port_on(interface):
+            self._port_on(interface)
+        command = 'interface vlan ' + vlan + '\r\n tagged ' + \
+            self.interface_type + ' ' + interface
+        self._execute(interface, CONFIG, command)
+
+    def _remove_vlan_from_trunk(self, interface, vlan):
+        """ Remove a vlan from a trunk port.
+
+        Args:
+            interface: interface to remove the vlan from
+            vlan: vlan to remove
+        """
+        command = 'interface vlan ' + vlan + '\r\n no tagged ' + \
+            self.interface_type + ' ' + interface
+        self._execute(interface, CONFIG, command)
+
+    def _remove_all_vlans_from_trunk(self, interface):
+        """ Remove all vlan from a trunk port.
+
+        Args:
+            interface: interface to remove the vlan from
+
+        this is kind slow for a number of reasons
+        1. we are using the REST API CLI
+        2. we have to go to every vlan and then remove our port from it;
+           that makes multiple calls.
+        """
```

Comment:
I don't think we should do it right now, but question: is it possible to parallelize this, or will the switch choke on that? I suspect we can improve on the latency if we have more than one request in flight at a time. Also, how slow are we talking?

Reply:
I think it's doable within this PR. 

This is what `_remove_native_vlan` does:

```python
        command = 'interface vlan ' + vlan + '\r\n no tagged ' + \
            self.interface_type + ' ' + interface
```
that translates to 2 commands in one call to the switch. The commands are separated by \r\n
```
switch(conf)# interface vlan vlanno
switch(conf-if-vl-vlanno)# no tagged gigabitethernet 1/3
```

So for removing multiple vlans, I should be able to generate a huge command (separated by \r\n) and send that in one call. That would probably be faster.

> Also, how slow are we talking?

Both deployment tests take around 340 seconds to complete.

Reply:
So, I did this. The deployment tests don't seem to be any faster. But I guess the speedup is proportional to part of program that can be parallelized.